hoomd/md/pytest/test_rigid.py

```diff
@@ -453,3 +453,160 @@ def test_rigid_dof(
+    # calculated in independent code.
+    # Expected velocity is tangential velocity plus center of mass velocity.
+    expected_tangential_velocity = [
+        [0, 0, 0],
+        [-6.25e-08, 5.00e-03, 0.00e00],
+        [6.25e-08, -5.00e-03, 0.00e00],
+        [5.00e-03, 6.25e-08, 0.00e00],
+        [-5.00e-03, -6.25e-08, 0.00e00],
+    ]
+    expected_velocity = np.add(expected_tangential_velocity, com_velocity)
+    new_snapshot = sim.state.get_snapshot()
+    if new_snapshot.communicator.rank == 0:
+        assert np.allclose(expected_velocity, new_snapshot.particles.velocity)
+
+
+@pytest.mark.parametrize(
+    "com_velocity",
+    [(0, 0, 0), (1, 2, 3)],
+)
+@pytest.mark.parametrize(
+    "omega",
+    [
+        [10, 0, 0],
+        [0, 10, 0],
+        [0, 0, 10],
+        [3, 4, 10],
+    ],
+)
+@pytest.mark.parametrize(
+    "body_definition,moment_of_inertia,mass",
+    [
+        (
+            {
+                "constituent_types": ["B", "B", "B", "B"],
+                "positions": [
+                    [1, 0, -1 / (2 ** (1.0 / 2.0))],
+                    [-1, 0, -1 / (2 ** (1.0 / 2.0))],
+                    [0, -1, 1 / (2 ** (1.0 / 2.0))],
+                    [0, 1, 1 / (2 ** (1.0 / 2.0))],
+                ],
+                "orientations": [(1.0, 0.0, 0.0, 0.0)] * 4,
+            },
+            [1, 1, 1],
+            [1],
+        ),
+        (
+            {
+                "constituent_types": ["B", "B"],
+                "positions": [[-1.2, 0, 0], [1.2, 0, 0]],
+                "orientations": [[1, 0, 0, 0], [1, 0, 0, 0]],
+            },
+            [0.0, 2.88, 2.88],
+            [2],
+        ),
+    ],
+    ids=["simple-inertia-body", "dimer-body"],
+)
+def test_velocity_constituents_constant_angmom(
+    com_velocity,
+    omega,
+    body_definition,
+    moment_of_inertia,
+    mass,
+    simulation_factory,
+    one_particle_snapshot_factory,
+):
+    """Test velocity of constituents updates correctly with 0 timestep."""
+    initial_snapshot = one_particle_snapshot_factory(particle_types=["A", "B"])
+    # steps to calculate angular momentum quaternion from desired angular
+    # velocity are typically the following:
+    # 1. Rotate space-frame angular velocity to body frame
+    # 2. Multiply by moment of inertia tensor in body frame
+    # 3. Promote to a quaternion with 0 real part and multiply by orientation
+    # 4. Multiply by 2
+    angmom = np.zeros(4)
+    angmom[1:] = 2 * np.multiply(omega, moment_of_inertia)
+    if initial_snapshot.communicator.rank == 0:
+        initial_snapshot.particles.moment_inertia[:] = [moment_of_inertia]
+        initial_snapshot.particles.angmom[:] = [angmom]
+        initial_snapshot.particles.mass[:] = mass
+        initial_snapshot.particles.velocity[:] = [com_velocity]
+
+    sim = simulation_factory(initial_snapshot)
+    sim.seed = 5
+
+    rigid = md.constrain.Rigid()
+    rigid.body["A"] = body_definition
+    rigid.create_bodies(sim.state)
```

Comment:
To test that the masses of constituent particles are retained correctly, I think you can take another snapshot here and assign the masses of the constituent particles. Then, you can test after that the run that they have the same values as what you have set.

Reply:
Ah, I was operating under the assumption that snapshot were copies of the simulation state and changes to them wouldn't actually change the simulation state.

Reply:
You are correct that a snapshot on its own is just a copy! However, you can use this method:
https://hoomd-blue.readthedocs.io/en/v5.1.0/hoomd/state.html#hoomd.State.set_snapshot
to restore a modified snapshot.